cli/cmd/upgrade.go

```diff
@@ -216,3 +219,10 @@ func fetchIssuer(k *kubernetes.Clientset, trustPEM string) (string, string, time
 
 	return keyPEM, crtPEM, crt.Certificate.NotAfter, nil
 }
+
+//prints the error message and quits the upgrade process
+func upgradeErrorf(format string, a ...interface{}) {
```

Comment:
This function causes the process to exit on all errors. I think we should preserve the previous way of error handling, where the error is propagated back to the framework.

We should be able to use `defer` to achieve what we want with something like:
```diff
diff --git a/cli/cmd/upgrade.go b/cli/cmd/upgrade.go
index a126ab17..1936fc0a 100644
--- a/cli/cmd/upgrade.go
+++ b/cli/cmd/upgrade.go
@@ -40,10 +40,20 @@ Note that the default flag values for this command come from the Linkerd control
 plane. The default values displayed in the Flags section below only apply to the
 install command.`,
 		RunE: func(cmd *cobra.Command, args []string) error {
+			upgradeSucceeded := false
+			defer func() {
+				if !upgradeSucceeded {
+					fmt.Fprintf(os.Stderr, "\n%s %s\n", failStatus, failMessage)
+					return
+				}
+
+				fmt.Fprintf(os.Stderr, "\n%s %s\n", okStatus, okMessage)
+			}()
+
 			// We need a Kubernetes client to fetch configs and issuer secrets.
 			k, err := options.newK8s()
 			if err != nil {
-				upgradeErrorf("failed to create a kubernetes client: %s", err)
+				return fmt.Errorf("failed to create a kubernetes client: %s", err)
 			}
 
 			// We fetch the configs directly from kubernetes because we need to be able
@@ -52,7 +62,7 @@ install command.`,
 			// control plane.
 			configs, err := fetchConfigs(k)
 			if err != nil {
-				upgradeErrorf("could not fetch configs from kubernetes: %s", err)
+				return fmt.Errorf("could not fetch configs from kubernetes: %s", err)
 			}
 
 			// If the install config needs to be repaired--either because it did not
@@ -80,21 +90,22 @@ install command.`,
 
 			values, err := options.buildValuesWithoutIdentity(configs)
 			if err != nil {
-				upgradeErrorf("could not build install configuration: %s", err)
+				return fmt.Errorf("could not build install configuration: %s", err)
 			}
 
 			identityValues, err := fetchIdentityValues(k, options.controllerReplicas, configs.GetGlobal().GetIdentityContext())
 			if err != nil {
-				upgradeErrorf("Unable to fetch the existing issuer credentials from Kubernetes.\nError: %s\n", err)
+				fmt.Fprintln(os.Stderr, "Unable to fetch the existing issuer credentials from Kubernetes.")
+				fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+				os.Exit(1)
 			}
 			values.Identity = identityValues
 
 			if err = values.render(os.Stdout, configs); err != nil {
-				upgradeErrorf("could not render install configuration: %s", err)
+				return fmt.Errorf("could not render install configuration: %s", err)
 			}
 
-			fmt.Fprintf(os.Stderr, "\n%s %s\n", okStatus, okMessage)
-
+			upgradeSucceeded = true
 			return nil
 		},
 	}
@@ -219,10 +230,3 @@ func fetchIssuer(k *kubernetes.Clientset, trustPEM string) (string, string, time
 
 	return keyPEM, crtPEM, crt.Certificate.NotAfter, nil
 }
-
-//prints the error message and quits the upgrade process
-func upgradeErrorf(format string, a ...interface{}) {
-	template := fmt.Sprintf("%s %s\n%s\n", failStatus, format, failMessage)
-	fmt.Fprintf(os.Stderr, template, a...)
-	os.Exit(1)
-}

```

Reply:
It was my suggestion to explicitly exit when printing "For troubleshooting help...", rather than returning an error, since in my mind, the link to the URL replaces the need for the standard cobra help output for the command, which prints all of the flags. Printing all of the flags also moves the URL that we print out of the terminal window, so I'd prefer if we did one or the other, but not both.

Reply:
> since in my mind, the link to the URL replaces the need for the standard cobra help output for the command, which prints all of the flags. Printing all of the flags also moves the URL that we print out of the terminal window

@scottcarol and I discussed about this earlier; definitely not a fan of the block of 'help' text in this case. If the way error is handled here can be different from the rest of the CLI, let's go with this approach of printing the URL, follow by an immediate exit. Since `upgrade` failure is a big deal, the error message and next step (i.e. go to the URL) shouldn't be mingled with the 'help' text.